src/python/graplctl/graplctl/swarm/commands.py

```diff
@@ -0,0 +1,323 @@
+@click.option(
+    "-i",
+    "--swarm-id",
+    type=click.STRING,
+    help="unique id for this swarm cluster (random default)",
+    default=str(uuid.uuid4()),
+)
+@pass_graplctl_state
+def swarm_create(
+    graplctl_state: GraplctlState,
+    num_managers: int,
+    num_workers: int,
+    instance_type: str,
+    swarm_id: str,
+) -> None:
+    docker_swarm_ops.create_swarm(
+        graplctl_state=graplctl_state,
+        num_managers=num_managers,
+        num_workers=num_workers,
+        instance_type=instance_type,
+        swarm_id=swarm_id,
+    )
+
+
+@swarm.command(help="list swarm ids for each of the swarm clusters", name="ls")
+@pass_graplctl_state
+def swarm_ls(graplctl_state: GraplctlState):
+    for swarm_id in docker_swarm_ops.swarm_ls(graplctl_state):
+        click.echo(swarm_id)
+
+
+@swarm.command(help="get instance ids for a docker swarm's managers", name="managers")
+@click.option(
+    "-i",
+    "--swarm-id",
+    type=click.STRING,
+    help="unique id of the swarm cluster",
+    required=True,
+)
+@pass_graplctl_state
+def swarm_managers(graplctl_state: GraplctlState, swarm_id: str):
+    for manager_instance in docker_swarm_ops.swarm_instances(
+        ec2=graplctl_state.ec2,
+        deployment_name=graplctl_state.grapl_deployment_name,
+        version=graplctl_state.grapl_version,
+        region=graplctl_state.grapl_region,
+        swarm_id=swarm_id,
+        swarm_manager=True,
+    ):
+        click.echo(manager_instance.instance_id)
+
+
+@swarm.command(help="terminate a docker swarm cluster's instances", name="destroy")
+@click.option(
+    "-i",
+    "--swarm-id",
+    type=click.STRING,
+    help="unique id of the swarm cluster",
+    required=True,
+)
+@click.confirmation_option(prompt="are you sure you want to destroy the swarm cluster?")
+@pass_graplctl_state
+def swarm_destroy(graplctl_state: GraplctlState, swarm_id: str):
+    click.echo(f"destroying swarm {swarm_id}")
+    docker_swarm_ops.destroy_swarm(graplctl_state=graplctl_state, swarm_id=swarm_id)
+    click.echo(f"destroyed swarm {swarm_id}")
+
+
+@swarm.command(help="execute a command on a swarm manager", name="exec")
+@click.option(
+    "-i",
+    "--swarm-id",
+    type=click.STRING,
+    help="unique id of the swarm cluster",
+    required=True,
+)
+@click.argument("command", nargs=-1, type=click.STRING)
+@pass_graplctl_state
+def swarm_exec(graplctl_state: GraplctlState, swarm_id: str, command: List[str]):
+    click.echo(
+        docker_swarm_ops.exec_(
+            ec2=graplctl_state.ec2,
+            ssm=graplctl_state.ssm,
+            deployment_name=graplctl_state.grapl_deployment_name,
+            region=graplctl_state.grapl_region,
+            version=graplctl_state.grapl_version,
+            swarm_id=swarm_id,
+            command=command,
+        )
+    )
+
+
+@swarm.command(help="scale up a docker swarm cluster", name="scale")
+@click.option(
+    "-m",
+    "--num-managers",
+    type=click.IntRange(min=0, max=None),
```

Comment:
Probably not $1M, because there aren't usually enough on-demand instances available in a region. The cap is probably in the thousands or tens of thousands. It's hard to imagine what a reasonable limit would be to set here that would (a) catch an error and (b) not interfere with legitimate usage.

Reply:
👍

Reply:
done